spacy/kb/candidate.py

```diff
@@ -0,0 +1,116 @@
+import abc
+from typing import List, Union, Callable
+
+from ..errors import Errors
+
+
+class Candidate(abc.ABC):
+    """A `Candidate` object refers to a textual mention that may or may not be resolved
+    to a specific `entity_id` from a Knowledge Base. This will be used as input for the entity_id linking
+    algorithm which will disambiguate the various candidates to the correct one.
+    Each candidate (mention, entity_id) pair is assigned a certain prior probability.
+
+    DOCS: https://spacy.io/api/kb/#candidate-init
+    """
+
+    def __init__(
+        self,
+        mention: str,
+        entity_id: Union[str, int],
+        entity_vector: List[float],
+        prior_prob: float,
+    ):
+        """Initializes properties of `Candidate` instance.
+        mention (str): Mention text for this candidate.
+        entity_id (Union[str, int]): Unique entity ID.
+        entity_vector (List[float]): Entity embedding.
+        prior_prob (float): Prior probability of entity for this mention - i.e. the probability that, independent of
+            the context, this mention resolves to this entity_id in the corpus used to build the knowledge base. In
+            cases in which this isn't always possible (e.g.: the corpus to analyse contains mentions that the KB corpus
+            doesn't) it might be better to eschew this information and always supply the same value.
+        """
+        self._mention = mention
+        self._entity_id = entity_id
+        # Note that hashing an int value yields the same int value.
+        self._entity_id_hash = hash(entity_id)
+        self._entity_vector = entity_vector
+        self._prior_prob = prior_prob
+
+    @property
+    def entity_id(self) -> Union[str, int]:
+        """RETURNS (Union[str, int]): Unique entity ID."""
+        return self._entity_id
+
+    @property
+    def entity_id_int(self) -> int:
+        """RETURNS (int): Numerical representation of entity ID (if entity ID is numerical, this is just the entity ID,
+        otherwise the hash of the entity ID string)."""
+        return self._entity_id_hash
+
+    @property
+    def entity_id_str(self) -> str:
+        """RETURNS (str): String representation of entity ID."""
+        return str(self._entity_id)
```

Comment:
Can't we use the same convention as what we do on `Token`, to have the underscored version return the `str` and the normal one the hash? And not have a method that returns the `Union`, I don't think we actually need it?

Reply:
Yeah, that's cleaner. Done in https://github.com/explosion/spaCy/pull/12341/commits/cea58ade8969592266483848feae3fd06f537e6b.

Reply:
It is a bit confusing though possibly that we ask to specify `entity_id: Union[str, int]` and then have a property `entity_id() -> int`. Perhaps we should rename the `__init__()` argument?

Reply:
Or we rename the properties `entity_id` and `entity_id_` to `entity` and `entity_` instead. Also not that nice.

Reply:
Hmmmm. I think the current solution is fine. It's not 100% ideal but I don't think any of the alternatives are better.